File: backend/app/services/answer_synthesizer.py
"""Answer synthesis service — generates grounded answers from evidence packs."""

import time
import json
import logging
from dataclasses import dataclass, field

from app.providers.interface import GenerateTextOptions
from app.providers.router import get_provider_router
from app.services.evidence_builder import EvidencePack

logger = logging.getLogger(__name__)

SYSTEM_PROMPT = """You are GitLore, a repository archaeology assistant. You explain WHY code exists by analyzing historical evidence from commits, PRs, issues, and documentation.

RULES:
1. Start with a direct answer to the question.
2. Cite specific evidence (commits, PRs, issues) with links when available.
3. Distinguish explicit evidence from inference.
4. If evidence is weak or missing, say so explicitly.
5. Never fabricate historical rationale.
6. If multiple explanations are plausible, present ranked hypotheses.
7. Be concise and factual.

FORMAT:
- Direct answer (1-3 sentences)
- Evidence trail (bullet points with citations)
- Confidence assessment
- Uncertainty notes if applicable"""


@dataclass
class SynthesisResult:
    answer_text: str
    confidence: float
    uncertainty_notes: str
    model_used: str
    latency_ms: int
    hypotheses: list[dict] = field(default_factory=list)


class AnswerSynthesizer:
    """Synthesizes grounded answers from evidence packs using LLM providers."""

    def __init__(self):
        self.router = get_provider_router()

    async def synthesize(
        self,
        query_text: str,
        evidence_pack: EvidencePack,
        model_override: str | None = None,
    ) -> SynthesisResult:
        """Generate a grounded answer from evidence."""
        start = time.monotonic()

        # Build the evidence context
        evidence_context = self._format_evidence(evidence_context=evidence_pack)

        user_prompt = f"""Question: {query_text}

Available Evidence ({evidence_pack.total_sources} sources, coverage: {evidence_pack.coverage}):

{evidence_context}

{self._format_gaps(evidence_pack.gaps)}

Based on this evidence, provide a grounded answer. Cite specific artifacts by title and type."""

        model = model_override or "groq:llama-3.1-8b-instant"

        try:
            result = await self.router.generate_text(
                GenerateTextOptions(
                    prompt=user_prompt,
                    system=SYSTEM_PROMPT,
                    temperature=0.2,
                    max_tokens=2048,
                ),
                model_override=model,
            )
            latency_ms = int((time.monotonic() - start) * 1000)

            # Calculate confidence from evidence quality
            confidence = self._compute_confidence(evidence_pack)

            # Extract uncertainty notes
            uncertainty = self._extract_uncertainty(result.text, evidence_pack)

            # Build hypotheses
            hypotheses = self._extract_hypotheses(result.text)

            return SynthesisResult(
                answer_text=result.text,
                confidence=confidence,
                uncertainty_notes=uncertainty,
                model_used=f"{result.provider}:{result.model}",
                latency_ms=latency_ms,
                hypotheses=hypotheses,
            )

        except Exception as e:
            logger.error("Answer synthesis failed: %s", e)
            latency_ms = int((time.monotonic() - start) * 1000)
            return SynthesisResult(
                answer_text=self._fallback_answer(query_text, evidence_pack),
                confidence=0.1,
                uncertainty_notes=f"LLM synthesis failed: {e}. Showing evidence-based fallback.",
                model_used="fallback",
                latency_ms=latency_ms,
            )

    def _format_evidence(self, evidence_context: EvidencePack) -> str:
        lines = []
        for i, entry in enumerate(evidence_context.entries, 1):
            lines.append(f"[{i}] ({entry.artifact_type.upper()}) {entry.artifact_title}")
            lines.append(f"    Claim: {entry.claim}")
            lines.append(f"    Excerpt: {entry.excerpt[:200]}")
            if entry.artifact_url:
                lines.append(f"    URL: {entry.artifact_url}")
            lines.append(f"    Relevance: {entry.relevance_score:.2f}")
            lines.append("")
        return "\n".join(lines)

    def _format_gaps(self, gaps: list[str]) -> str:
        if not gaps:
            return ""
        return "\nGaps in evidence:\n" + "\n".join(f"- {g}" for g in gaps)

    def _compute_confidence(self, pack: EvidencePack) -> float:
        if pack.coverage == "sufficient":
            base = 0.7
        elif pack.coverage == "partial":
            base = 0.4
        else:
            base = 0.15

        direct_count = sum(1 for e in pack.entries if e.is_direct)
        high_rel = sum(1 for e in pack.entries if e.relevance_score >= 0.5)
        return min(base + direct_count * 0.05 + high_rel * 0.03, 0.95)

    def _extract_uncertainty(self, answer: str, pack: EvidencePack) -> str:
        notes = []
        if pack.coverage == "insufficient":
            notes.append("Evidence coverage is insufficient — answer may be incomplete.")
        if pack.coverage == "partial":
            notes.append("Some evidence gaps exist.")
        if len(pack.entries) < 3:
            notes.append("Limited sources available for this answer.")
        return " ".join(notes) if notes else ""

    def _extract_hypotheses(self, answer: str) -> list[dict]:
        hypotheses = []
        lines = answer.split("\n")
        rank = 1
        for line in lines:
            line = line.strip()
            if line.startswith("- ") or line.startswith("* ") or (len(line) > 10 and ":" in line):
                text = line.lstrip("-* ").strip()
                if len(text) > 20:
                    hypotheses.append({
                        "rank": rank,
                        "explanation": text[:300],
                        "confidence": max(0.3, 0.8 - rank * 0.15),
                        "evidence_ids": [],
                    })
                    rank += 1
                    if rank > 3:
                        break
        return hypotheses

    def _fallback_answer(self, query: str, pack: EvidencePack) -> str:
        if not pack.entries:
            return (
                f"**Unable to provide a grounded answer.**\n\n"
                f"No relevant evidence was found for: {query}\n\n"
                f"This could mean the repository hasn't been fully indexed, "
                f"or the query targets code/history not yet captured."
            )

        lines = [
            f"Based on available evidence ({len(pack.entries)} sources, {pack.coverage} coverage):\n",
        ]
        for i, entry in enumerate(pack.entries[:5], 1):
            lines.append(f"**{i}. {entry.artifact_title}** ({entry.artifact_type})")
            lines.append(f"   {entry.claim}")
            lines.append(f"   Excerpt: {entry.excerpt[:150]}")
            if entry.artifact_url:
                lines.append(f"   Source: {entry.artifact_url}")
            lines.append("")

        lines.append("*Note: This is an evidence-based summary, not an LLM-synthesized answer.*")
        return "\n".join(lines)

File: src/lib/services/answer-synthesizer.ts
import * as types from '@/types';
import { getRouter } from '@/lib/providers/router';
import { QueryInterpreter } from '@/lib/pipeline/query-interpreter';

/**
 * AnswerSynthesizer generates grounded, evidence-based answers.
 * It NEVER invents historical rationale and always cites sources.
 */
export class AnswerSynthesizer {
  async synthesize(input: types.SynthesisInput): Promise<{
    answerText: string;
    confidence: number;
    uncertaintyNotes: string;
    hypotheses: types.Hypothesis[];
    modelUsed: string;
    latencyMs: number;
  }> {
    const start = Date.now();
    const router = getRouter();
    const evidenceContext = this.buildEvidenceContext(input.evidence);
    const isDiagram = QueryInterpreter.isDiagramQuery(input.originalQuestion);

    const systemPrompt = isDiagram
      ? `You are an expert software architect analyzing a codebase. Generate Mermaid diagrams to visualize the system.

RULES:
1. Analyze the evidence to understand the project structure, modules, and relationships.
2. Generate a valid Mermaid diagram that represents the system architecture/data flow/component structure.
3. Use appropriate diagram type: flowchart TB for architecture, flowchart LR for data flow, graph TD for dependencies, sequenceDiagram for interactions.
4. Add meaningful labels and descriptions to each node.
5. After the diagram, provide a brief explanation of what it shows.
6. Always wrap the Mermaid code in \`\`\`mermaid code fences.

Format your answer as:
**System Architecture:**

\`\`\`mermaid
[mermaid diagram code]
\`\`\`

**Explanation:** [brief explanation of the diagram]`
      : `You are an expert code archaeologist analyzing repository history. 
Your role: Explain why code exists the way it does, citing historical evidence.

RULES:
1. Only make claims supported by provided evidence. If insufficient, say so.
2. Cite every key claim with [TYPE: Title].
3. Distinguish: [EVIDENCE] = directly stated, [INFERRED] = logical conclusion, [MISSING] = not found.
4. Never invent commit messages, PR descriptions, or author intent.
5. If multiple explanations are plausible, present ranked hypotheses.
6. Be honest about uncertainty.
7. Prefer explicit discussion over inferred intent.`;

    const diagramInstructions = isDiagram
      ? `Based on the evidence (source files, commits, PRs, docs), generate a Mermaid diagram that visualizes the project's architecture, module relationships, or data flow. The diagram should be accurate to the codebase structure.`
      : '';

    const userPrompt = `## Query\n${input.originalQuestion}\n\n## Query Context\nIntent: ${input.query.intent}\nEntities: ${JSON.stringify(input.query.entities)}\nTime: ${JSON.stringify(input.query.timeHints)}\n\n## Evidence Pack\nCoverage: ${input.evidence.coverage}\nGaps: ${input.evidence.gaps.join('\n')}\n\n${evidenceContext}\n\n## Instructions\n${isDiagram ? diagramInstructions : `Write a concise answer that:
1. Directly answers the question
2. Cites evidence for each claim using [TYPE: Title]
3. Includes confidence assessment
4. Notes what is unknown
5. Includes timeline context where relevant
6. If insufficient evidence, state "Insufficient evidence" and show closest trail

Format:
**Answer:** [answer]
**Confidence:** [high/medium/low - explain why]
**Key Evidence:** [list with citations]
**Uncertainties:** [what's unknown]
**Timeline:** [key chronological events]`}`;


    try {
      const result = await router.generateText(
        { prompt: userPrompt, system: systemPrompt, temperature: 0.2, maxTokens: 2048 },
        'answerSynthesis'
      );
      const latency = Date.now() - start;
      return {
        answerText: result.text,
        confidence: this.extractConfidence(result.text),
        uncertaintyNotes: this.extractUncertainty(result.text),
        hypotheses: this.extractHypotheses(result.text),
        modelUsed: `${result.provider}/${result.model}`,
        latencyMs: latency,
      };
    } catch (err) {
      console.error(`[Synthesizer] LLM failed:`, err instanceof Error ? err.message : err);
      return {
        answerText: this.generateFallbackAnswer(input),
        confidence: 0.3,
        uncertaintyNotes: 'LLM synthesis failed. Evidence summary provided.',
        hypotheses: [],
        modelUsed: 'fallback',
        latencyMs: Date.now() - start,
      };
    }
  }

  private buildEvidenceContext(evidence: types.EvidencePack): string {
    return evidence.entries.map((e, i) =>
      `--- Source ${i + 1} ---\nType: ${e.artifact.artifact_type.toUpperCase()}\nTitle: ${e.artifact.title || 'Untitled'}\nAuthor: ${e.artifact.author || 'Unknown'}\nDate: ${e.artifact.date ? new Date(e.artifact.date).toISOString().split('T')[0] : 'Unknown'}\nURL: ${e.artifact.url || 'N/A'}\nRelevance: ${(e.relevanceScore * 100).toFixed(0)}%\nExcerpt: ${e.excerpt}`
    ).join('\n');
  }

  private extractConfidence(text: string): number {
    const lower = text.toLowerCase();
    if (lower.includes('confidence: high')) return 0.85;
    if (lower.includes('confidence: medium')) return 0.6;
    if (lower.includes('confidence: low')) return 0.3;
    if (lower.includes('insufficient evidence')) return 0.15;
    const citations = text.match(/\[.*?\]/g);
    return citations && citations.length >= 2 ? 0.6 : 0.4;
  }

  private extractUncertainty(text: string): string {
    const m = text.match(/\*\*uncertainties?\*\*:?\s*([^\n]*)/i);
    if (m) return m[1].trim();
    const m2 = text.match(/\[missing\].*?(?=\n|$)/gi);
    return m2 ? m2.join('; ') : '';
  }

  private extractHypotheses(text: string): types.Hypothesis[] {
    const h: types.Hypothesis[] = [];
    const h1 = text.match(/\*\*Hypothesis 1?\*\*:?\s*([^\n]*)/i);
    const h2 = text.match(/\*\*Hypothesis 2?\*\*:?\s*([^\n]*)/i);
    if (h1) h.push({ rank: 1, explanation: h1[1].trim(), confidence: 0.5, evidence_ids: [] });
    if (h2) h.push({ rank: 2, explanation: h2[1].trim(), confidence: 0.3, evidence_ids: [] });
    return h;
  }

  private generateFallbackAnswer(input: types.SynthesisInput): string {
    const ev = input.evidence;
    if (ev.entries.length === 0) {
      return '**Insufficient evidence found to answer this question.**\n\nThe repository does not contain artifacts that directly address this query. Consider rephrasing or checking if the repository has been fully indexed.';
    }

    const lines: string[] = [];
    lines.push(`Based on ${ev.entries.length} sources from repository history:\n`);

    // Group by type
    const byType: Record<string, typeof ev.entries> = {};
    for (const entry of ev.entries) {
      const type = entry.artifact.artifact_type;
      if (!byType[type]) byType[type] = [];
      byType[type].push(entry);
    }

    for (const [type, entries] of Object.entries(byType)) {
      const label = type === 'doc' ? 'Documentation' : type === 'pr' ? 'Pull Requests' : type === 'issue' ? 'Issues' : type === 'commit' ? 'Commits' : type.toUpperCase();
      lines.push(`**${label} (${entries.length}):**`);
      for (const entry of entries.slice(0, 3)) {
        const title = entry.artifact.title || 'Untitled';
        const author = entry.artifact.author ? ` by ${entry.artifact.author}` : '';
        const date = entry.artifact.date ? ` (${new Date(entry.artifact.date).toLocaleDateString()})` : '';
        // Show a cleaned excerpt (first 300 chars, skip JSON noise)
        let excerpt = entry.excerpt || '';
        // Try to extract readable text from JSON content
        try {
          const parsed = JSON.parse(excerpt);
          if (parsed.message) excerpt = parsed.message;
          else if (parsed.query) excerpt = `Query: ${parsed.query}`;
        } catch { /* not JSON, use as-is */ }
        excerpt = excerpt.replace(/```json[\s\S]*?```/g, '').replace(/```[\s\S]*?```/g, '').substring(0, 300).trim();
        if (excerpt) {
          lines.push(`- **${title}**${author}${date}\n  ${excerpt}`);
        } else {
          lines.push(`- **${title}**${author}${date}`);
        }
      }
    }

    if (ev.gaps.length > 0) {
      lines.push(`\n**Note:** ${ev.gaps[0]}`);
    }
    return lines.join('\n');
  }
}
